content/threads_intro.tex

```diff
@@ -0,0 +1,67 @@
+This section specifies the interaction between \openshmem interfaces and the
+user threads, and also describes the routines that can be used for initializing and 
+querying the thread environment. There are four levels of threading supported by
+the \openshmem implementation.
+ 
+\begin{itemize}
+\item {\bf \CONST{SHMEM\_THREAD\_SINGLE}} The \openshmem program may not be multithreaded. 
+
+\item {\bf \CONST{SHMEM\_THREAD\_FUNNELED}}
+The \openshmem program may be multithreaded. However, the 
+program must ensure that only the main thread invokes the \openshmem
+interfaces. The main thread is the thread that invokes either \FUNC{shmem\_init} or
+\FUNC{shmem\_init\_thread}.
+
+\item {\bf \CONST{SHMEM\_THREAD\_SERIALIZED}} 
+The \openshmem program may be multithreaded. However, the 
+program must ensure that the \openshmem interfaces 
+are not invoked concurrently by multiple threads.
+
+\item {\bf \CONST{SHMEM\_THREAD\_MULTIPLE}}
+The \openshmem program may be multithreaded and any 
+thread may invoke the \openshmem interfaces.
+\end{itemize}
+
+\hspace{-15pt}
+{\bf Semantics}
+
+\begin{itemize}
+\item
+In \CONST{SHMEM\_THREAD\_FUNNELED}, \CONST{SHMEM\_THREAD\_SERIALIZED}, and \CONST{SHMEM\_THREAD\_MULTIPLE} thread
+levels, the \FUNC{shmem\_init} and \FUNC{shmem\_finalize} calls may only be invoked by the same thread.
```

Comment:
shmem_init_thread

Reply:
shmem_init is correct. I haven't introduced shmem_init_thread interface at this point, so I did not wanted to introduce in the text.

Reply:
I agree with @abouteiller, shmem_init_thread makes more sense here (actually, both init and init_thread make sense, so you could list both).  The forward reference should be fine since this section introduces the new routine.

Reply:
I don't have a strong opinion here, but in our previous meetings many members disliked the forward references. :)

Option 1 : Rephrase the sentence 

 _"..levels, the library initialization (shmem_init, shmem_init_thread) and shmem_finalize calls may only be invoked by the same thread."_

 

Option 2: Rephrase the sentence 

_"..levels, the library initialization (shmem_init, shmem_init_thread) and finalization (shmem_finalize) calls may only be invoked by the same thread."_

_

Option 3 : Alternatively, I can add a sentence in shmem_init_thread.